iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp

```diff
@@ -0,0 +1,357 @@
+      return false;
+    uses.push_back(userOp);
+  }
+  return true;
+}
+
+void MemRefUsageAnalysis::analyzePlaceholder(
+    IREE::PlaceholderOp placeholderOp) {
+  SmallVector<Operation *, 4> vectorUses;
+  MemRefType memrefType = placeholderOp.getType().dyn_cast<MemRefType>();
+  if (memrefType && isMemRefAndVectorizable(memrefType) &&
+      allVectorUses(placeholderOp, vectorUses)) {
+    vectorize.insert(placeholderOp);
+    transferOps.insert(vectorUses.begin(), vectorUses.end());
+  }
+}
+
+void MemRefUsageAnalysis::analyzeAlloc(AllocOp allocOp) {
+  SmallVector<Operation *, 4> vectorUses;
+  if (isMemRefAndVectorizable(allocOp.getType()) &&
+      allVectorUses(allocOp, vectorUses)) {
+    vectorize.insert(allocOp);
+    transferOps.insert(vectorUses.begin(), vectorUses.end());
+  }
+}
+
+template <typename OpTy>
+class MemRefConversionPattern : public OpConversionPattern<OpTy> {
+ public:
+  MemRefConversionPattern<OpTy>(MLIRContext *context,
+                                const MemRefUsageAnalysis &memrefUsageAnalysis)
+      : OpConversionPattern<OpTy>::OpConversionPattern(context),
+        memrefUsageAnalysis(memrefUsageAnalysis) {}
+
+ protected:
+  const MemRefUsageAnalysis &memrefUsageAnalysis;
+};
+
+class ProcessFuncArg final : public MemRefConversionPattern<FuncOp> {
+ public:
+  using MemRefConversionPattern<FuncOp>::MemRefConversionPattern;
+  LogicalResult matchAndRewrite(
+      FuncOp funcOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override;
+};
+
+class ProcessTransferRead final
+    : public MemRefConversionPattern<vector::TransferReadOp> {
+ public:
+  using MemRefConversionPattern<
+      vector::TransferReadOp>::MemRefConversionPattern;
+  LogicalResult matchAndRewrite(
+      vector::TransferReadOp read, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    if (!memrefUsageAnalysis.transferConvert(read)) return failure();
+    vector::TransferReadOp::Adaptor adaptor(operands);
+    Value memref = adaptor.memref();
+    Location loc = read.getLoc();
+    Optional<unsigned> vecMemrefElemSize =
+        getBitWidth(memref.getType().cast<MemRefType>().getElementType());
+    Optional<unsigned> readElemSize =
+        getBitWidth(read.getMemRefType().getElementType());
+    Optional<unsigned> readVecSize = getBitWidth(read.getVectorType());
+    if (!vecMemrefElemSize || !readElemSize || !readVecSize) return failure();
+    unsigned ratio = *vecMemrefElemSize / *readElemSize;
+    SmallVector<Value, 4> indices(adaptor.indices().begin(),
+                                  adaptor.indices().end());
+    indices.back() = rewriter.create<SignedDivIOp>(
+        loc, indices.back(), rewriter.create<ConstantIndexOp>(loc, ratio));
+    // If the transfer_read can be replaced by a load after vectorization use
+    // LoadOp and cast back to the original type.
+    if (*vecMemrefElemSize == *readVecSize) {
```

Comment:
I am not convinced we should do this here. This seems like something that is better fit with the lowering of `vector.transfer_read` to `spv.load` operation.

If I understand this right, this is taking
```
vector.transfer_read %0[%c0, %c0], %cst : memref<4096x4096xf32>, vector<1x4xf32>
```

and converting this to

```
vector.transfer_read %0[%c0, %c0], %cst : memref<4096x1024x<vector<4xf32>>, vector<1x4xf32>
```

and doing the additional step of lowering this to
```
loadOp %0[%c0, %c0/4] : vector<1x4xf32>
```

I think I understand why this is done. This is because `vector.transfer_read` in `VectorToSPIRV` is lowered to the co-operative matrix multiply loads. To me it seems cleaner to make that decision while lowering from Vector dialect to SPIR-V dialect. You can have multiple patterns, one that lowers to a `spv.LoadOp`, and other that lowers to a `spv.CooperativeMatLoadNV` based on uses.

I am advocating for this cause this logic does not seem limited to IREE, and putting in MLIR might be more useful.

Reply:
> If I understand this right, this is taking
> 
> ```
> vector.transfer_read %0[%c0, %c0], %cst : memref<4096x4096xf32>, vector<1x4xf32>
> ```
> 
> and converting this to
> 
> ```
> vector.transfer_read %0[%c0, %c0], %cst : memref<4096x1024x<vector<4xf32>>, vector<1x4xf32>
> ```
> 
> and doing the additional step of lowering this to
> 
> ```
> loadOp %0[%c0, %c0/4] : vector<1x4xf32>
> ```

It actually transforms it to:
```
%1 = loadOp %0[%c0, %c0/4] : vector<4xf32>
shape_cast %1 to vector<1x4xf32>
```



> I think I understand why this is done. This is because `vector.transfer_read` in `VectorToSPIRV` is lowered to the co-operative matrix multiply loads.

This is not really the main reason. Right now convert to spv wouldn't convert it to cooperative matrix load since the cooperative matrix analysis would figure it shouldn't be a cooperative matrix type. The main reason I was doing it here was because the transformation can be expressed in standard/vector dialect and it seemed better to decouple it from the conversion to SPIR-V. I would imagine at some point there would be more complex cases we would want to lower to to several loads or handle the masked transfer_read by adding `scf.if` kind of operation. I'm afraid doing all that in conversion from standard to spv would make things too complicated.

> I am advocating for this cause this logic does not seem limited to IREE, and putting in MLIR might be more useful.

That's a good point. In my mind the whole pass was going to end up in MLIR but I realize now because of IREE::Placeholder this is not going to be possible. I wonder if we need an extra pass to map the transfer read/write to operations that can be lowered 1 to 1 to spv ops.

Reply:
> That's a good point. In my mind the whole pass was going to end up in MLIR but I realize now because of IREE::Placeholder this is not going to be possible. I wonder if we need an extra pass to map the transfer read/write to operations that can be lowered 1 to 1 to spv ops.

I dont know if we need a separate pass. I think we can just add Vector -> SPIR-V patterns and add it there. I was under the impression such a pass already exists, and thats where the `vector.contract` to `spv.CooperativeMatrixMulAddNV` transformation exists. I couldnt find it though. I think we should have `vector` to `spirv` patterns and then we can just add it to `ConvertToSPIRV` pass.

Reply:
> I dont know if we need a separate pass. I think we can just add Vector -> SPIR-V patterns and add it there. I was under the impression such a pass already exists, and thats where the `vector.contract` to `spv.CooperativeMatrixMulAddNV` transformation exists. I couldnt find it though. I think we should have `vector` to `spirv` patterns and then we can just add it to `ConvertToSPIRV` pass.

This is currently part of the IREE ConvertToSPIRV pass but to be honest the only reason why I have it in this pass is because there is no other way to represent CooperativeMatrix in GPU dialect and spv conversion needs to happen all at once (since spv has its own types). My understanding was that we would want to move that out of the ConvertToSPIRV pass.

I can move some of this logic in ConverToSPIRV, however the decision to insert the bitcast/shapecast still needs to be taken in this pass so that we have a chance to clean it up before going to SPIR-V. Since transfer_read/write allows the load type to be different from the vector type to match cooperative matrix load semantic but Load won't allow it I kind of need to know already what the operation will be lowered to. Maybe we can chat about it offline.

Reply:
Based on offline conversation, these will eventually move into MLIR core. So landing it as is for now is fine.